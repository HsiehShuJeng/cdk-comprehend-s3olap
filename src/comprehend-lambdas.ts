import * as sam from '@aws-cdk/aws-sam';
import * as cdk from '@aws-cdk/core';

export interface AccessConrtolLambdaProps {
  /**
   * The version of the serverless application.
   *
   * @default '1.0.2'
   */
  readonly semanticVersion?: string;
  /**
   * The minimum prediction confidence score above which PII classification and detection would be considered as final answer. Valid range (0.5 to 1.0).
   *
   * @default '0.5'
   */
  readonly confidenceThreshold?: string;
  /**
   * Number of threads to use for calling Comprehend's ContainsPiiEntities API. This controls the number of simultaneous calls that will be made from this Lambda.
   *
   * @default '20'
   */
  readonly containsPiiEntitiesThreadCount?: string;
  /**
   * Default language of the text to be processed. This code will be used for interacting with Comprehend.
   *
   * @default 'en'
   */
  readonly defaultLanguageCode?: string;
  /**
   * Default maximum document size (in bytes) that this function can process otherwise will throw exception for too large document size.
   *
   * @default '102400'
   */
  readonly documentMaxSize?: string;
  /**
   * Maximum document size (in bytes) to be used for making calls to Comprehend's ContainsPiiEntities API.
   *
   * @default '50000'
   */
  readonly documentMaxSizeContainsPiiEntities?: string;
  /**
   * Whether to support partial objects or not. Accessing partial object through http headers such byte-range can corrupt the object and/or affect PII detection accuracy.
   *
   * @default 'false'
   */
  readonly isPartialObjectSupported?: string;
  /**
   * Log level for Lambda function logging, e.g., ERROR, INFO, DEBUG, etc.
   *
   * @default 'INFO'
   */
  readonly logLevel?: string;
  /**
   * Maximum characters to overlap among segments of a document in case chunking is needed because of maximum document size limit.
   *
   * @default '200'
   */
  readonly maxCharsOverlap?: string;
  /**
   * List of comma separated PII entity types to be considered for access control. Refer Comprehend's documentation page for list of supported PII entity types.
   *
   * @default 'ALL'
   */
  readonly piiEntityTypes?: string;
  /**
   * True if publish metrics to Cloudwatch, false otherwise. See README.md for details on CloudWatch metrics.
   *
   * @default 'true'
   */
  readonly publishCloudWatchMetrics?: string;
  /**
   * Number of tokens/words to overlap among segments of a document in case chunking is needed because of maximum document size limit.
   *
   * @default '20'
   */
  readonly subsegmentOverlappingTokens?: string;
  /**
   * Handling logic for Unsupported files. Valid values are PASS and FAIL.
   *
   * @default 'FAIL'
   */
  readonly unsupportedFileHandling?: string;
}

export class AccessConrtolLambda extends cdk.Construct {
  /**
   * The name of the underlying resoure in the serverless application.
   *
   * @attribute
   */
  public readonly stackName: string;
  // readonly semanticVersion: string;
  // readonly confidenceThreshold: string;
  // readonly containsPiiEntitiesThreadCount: string;
  // readonly defaultLanguageCode: string;
  // readonly documentMaxSize: string;
  // readonly documentMaxSizeContainsPiiEntities: string;
  // readonly isPartialObjectSupported: string;
  // readonly logLevel: string;
  // readonly maxCharsOverlap: string;
  // readonly piiEntityTypes: string;
  // readonly publishCloudWatchMetrics: string;
  // readonly subsegmentOverlappingTokens: string;
  // readonly unsupportedFileHandling: string;
  /**
   * The entity of the serverless application.
   */
  private readonly samApplication: sam.CfnApplication;
  constructor(scope: cdk.Construct, id: string, props: AccessConrtolLambdaProps) {
    super(scope, id);
    const semanticVersion = props.semanticVersion ?? '1.0.2';
    const confidenceThreshold = props.confidenceThreshold ?? '0.5';
    const containsPiiEntitiesThreadCount = props.containsPiiEntitiesThreadCount ?? '20';
    const defaultLanguageCode = props.defaultLanguageCode ?? 'en';
    const documentMaxSize = props.documentMaxSize ?? '102400';
    const documentMaxSizeContainsPiiEntities = props.documentMaxSizeContainsPiiEntities ?? '50000';
    const isPartialObjectSupported = props.isPartialObjectSupported ?? 'false';
    const logLevel = props.logLevel ?? 'INFO';
    const maxCharsOverlap = props.maxCharsOverlap ?? '200';
    const piiEntityTypes = props.piiEntityTypes ?? 'ALL';
    const publishCloudWatchMetrics = props.publishCloudWatchMetrics ?? 'true';
    const subsegmentOverlappingTokens = props.subsegmentOverlappingTokens ?? '20';
    const unsupportedFileHandling = props.unsupportedFileHandling ?? 'FAIL';
    this.samApplication = new sam.CfnApplication(this, 'ComprehendPiiAccessControlS3ObjectLambda', {
      location: {
        applicationId: 'arn:aws:serverlessrepo:us-east-1:839782855223:applications/ComprehendPiiAccessControlS3ObjectLambda',
        semanticVersion: semanticVersion,
      },
      parameters: {
        ['ConfidenceThreshold']: confidenceThreshold,
        ['ContainsPiiEntitiesThreadCount']: containsPiiEntitiesThreadCount,
        ['DefaultLanguageCode']: defaultLanguageCode,
        ['DocumentMaxSize']: documentMaxSize,
        ['DocumentMaxSizeContainsPiiEntities']: documentMaxSizeContainsPiiEntities,
        ['IsPartialObjectSupported']: isPartialObjectSupported,
        ['LogLevel']: logLevel,
        ['MaxCharsOverlap']: maxCharsOverlap,
        ['PiiEntityTypes']: piiEntityTypes,
        ['PublishCloudWatchMetrics']: publishCloudWatchMetrics,
        ['SubsegmentOverlappingTokens']: subsegmentOverlappingTokens,
        ['UnsupportedFileHandling']: unsupportedFileHandling,
      },
    });
    this.stackName = cdk.Token.asString(this.samApplication.ref);
    for (var child of this.samApplication.node.children) {
      console.log(child.node.id);
      console.log(child.node.path);
    }
  }
}
