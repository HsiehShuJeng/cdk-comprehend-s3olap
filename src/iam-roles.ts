import * as iam from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';

export interface GeneralRoleProps {
  /**
   * The name of the IAM policy for the IAM role.
   *
   * @default 'general-role-s3olap-policy'
   */
  readonly policyName?: string;
  /**
     * The name of the object Lambda access point, which will be the same as the S3 acceess point for the S3 bucket in the demostration.
     *
     * @default 'accessctl-s3olap-survey-results-unknown-pii'
     */
  readonly objectLambdaAccessPointName?: string;
  /**
     * The name of the IAM role.
     *
     * @default 'GeneralRole'
     */
  readonly iamRoleName?: string;
}

/**
 * The role that you are going to assume (switch role)
 *
 * Explores how the S3 Object Lambda works.
 */
export class GeneralRole extends cdk.Construct {
  /**
     * The ARN of the IAM role.
     *
     */
  public readonly roleArn: string;
  /**
     * The name of the IAM role.
     *
     */
  public readonly roleName: string;
  /**
     * The unique string identifying the role.
     *
     */
  public readonly roleId: string;
  constructor(scope: cdk.Construct, id: string, props: GeneralRoleProps) {
    super(scope, id);
    const policyName = props.policyName ?? 'general-role-s3olap-policy';
    const objectLambdaAccessPointName = props.objectLambdaAccessPointName ?? 'accessctl-s3olap-survey-results-unknown-pii';
    const iamRoleName = props.iamRoleName ?? 'GeneralRole';
    const generalRole = new iam.Role(this, 'IamRole', {
      roleName: iamRoleName,
      assumedBy: new iam.AccountRootPrincipal(),
      inlinePolicies: {
        [policyName]: new iam.PolicyDocument({
          assignSids: true,
          statements: [
            new iam.PolicyStatement({
              sid: 'AllowListingObjects',
              effect: iam.Effect.ALLOW,
              actions: ['s3:ListBucket'],
              resources: ['*'],
            }),
            new iam.PolicyStatement({
              sid: 'AllowListingBucketsAndAccessPoints',
              effect: iam.Effect.ALLOW,
              actions: [
                's3:GetAccessPointForObjectLambda',
                's3:GetAccessPointConfigurationForObjectLambda',
                's3:ListAccessPointsForObjectLambda',
                's3:ListAllMyBuckets',
                's3:ListAccessPoints',
                's3:GetAccessPoint',
                's3:GetAccountPublicAccessBlock',
                's3:GetBucketPublicAccessBlock',
                's3:GetBucketPolicyStatus',
                's3:GetBucketAcl',
                's3:GetAccessPointPolicyStatus',
              ],
              resources: ['*'],
            }),
            new iam.PolicyStatement({
              sid: 'AllowObjectLambdaAccess',
              effect: iam.Effect.ALLOW,
              actions: [
                's3-object-lambda:Get*',
                's3-object-lambda:List*',
              ],
              resources: [`arn:${cdk.Aws.PARTITION}:s3-object-lambda:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:accesspoint/${objectLambdaAccessPointName}`],
            }),
            new iam.PolicyStatement({
              sid: 'AllowStandardAccessPointAccess',
              effect: iam.Effect.ALLOW,
              actions: [
                's3:Get*',
                's3:List*',
              ],
              resources: ['*'],
              conditions: {
                ['ForAnyValue:StringEquals']: {
                  'aws:CalledVia': ['s3-object-lambda.amazonaws.com'],
                },
              },
            }),
            new iam.PolicyStatement({
              sid: 'AllowLambdaInvocation',
              effect: iam.Effect.ALLOW,
              actions: [
                'lambda:InvokeFunction',
              ],
              resources: ['*'],
              conditions: {
                ['ForAnyValue:StringEquals']: {
                  'aws:CalledVia': ['s3-object-lambda.amazonaws.com'],
                },
              },
            }),
          ],
        }),
      },
    });
    this.roleArn = generalRole.roleArn;
    this.roleName = generalRole.roleName;
    this.roleId = generalRole.roleId;
  }
}


export interface AdminRoleProps {
  /**
   * The name of the IAM policy for the IAM role.
   *
   * @default 'admin-role-s3olap-policy'
   */
  readonly policyName?: string;
  /**
   * The name of the object Lambda access point, which will be the same as the S3 acceess point for the S3 bucket in the demostration.
   *
   * @default 'admin-s3olap-call-transcripts-known-pii'
   */
  readonly objectLambdaAccessPointName?: string;
  /**
   * The name of the IAM role.
   *
   * @default 'RedactionAdminRole'
   */
  readonly iamRoleName?: string;
}

export class AdminRole extends cdk.Construct {
  /**
   * The ARN of the IAM role.
   */
  public readonly roleArn: string;
  /**
   * The name of the IAM role.
   */
  public readonly roleName: string;
  /**
   * The unique string identifying the role.
   */
  public readonly roleId: string;
  constructor(scope: cdk.Construct, id: string, props?: AdminRoleProps) {
    super(scope, id);
    const policyName = props?.policyName ?? 'admin-role-s3olap-policy';
    const objectLambdaAccessPointName = props?.objectLambdaAccessPointName ?? 'admin-s3olap-call-transcripts-known-pii';
    const iamRoleName = props?.iamRoleName ?? 'RedactionAdminRole';
    const adminRole = new iam.Role(this, 'RedactionAdminRole', {
      roleName: iamRoleName,
      assumedBy: new iam.AccountRootPrincipal(),
      inlinePolicies: {
        [policyName]: new iam.PolicyDocument({
          assignSids: true,
          statements: [
            new iam.PolicyStatement({
              sid: 'AllowListingObjects',
              effect: iam.Effect.ALLOW,
              actions: ['s3:ListBucket'],
              resources: ['*'],
            }),
            new iam.PolicyStatement({
              sid: 'AllowListingBucketsAndAccessPoints',
              effect: iam.Effect.ALLOW,
              actions: [
                's3:GetAccessPointForObjectLambda',
                's3:GetAccessPointConfigurationForObjectLambda',
                's3:ListAccessPointsForObjectLambda',
                's3:ListAllMyBuckets',
                's3:ListAccessPoints',
                's3:GetAccessPoint',
                's3:GetAccountPublicAccessBlock',
                's3:GetBucketPublicAccessBlock',
                's3:GetBucketPolicyStatus',
                's3:GetBucketAcl',
                's3:GetAccessPointPolicyStatus',
              ],
              resources: ['*'],
            }),
            new iam.PolicyStatement({
              sid: 'AllowObjectLambdaAccess',
              effect: iam.Effect.ALLOW,
              actions: [
                's3-object-lambda:Get*',
                's3-object-lambda:List*',
              ],
              resources: [`arn:${cdk.Aws.PARTITION}:s3-object-lambda:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:accesspoint/${objectLambdaAccessPointName}`],
            }),
            new iam.PolicyStatement({
              sid: 'AllowStandardAccessPointAccess',
              effect: iam.Effect.ALLOW,
              actions: [
                's3:Get*',
                's3:List*',
              ],
              resources: ['*'],
              conditions: {
                ['ForAnyValue:StringEquals']: {
                  'aws:CalledVia': ['s3-object-lambda.amazonaws.com'],
                },
              },
            }),
            new iam.PolicyStatement({
              sid: 'AllowLambdaInvocation',
              effect: iam.Effect.ALLOW,
              actions: [
                'lambda:InvokeFunction',
              ],
              resources: ['*'],
              conditions: {
                ['ForAnyValue:StringEquals']: {
                  'aws:CalledVia': ['s3-object-lambda.amazonaws.com'],
                },
              },
            }),
          ],
        }),
      },
    });
    this.roleArn = adminRole.roleArn;
    this.roleName = adminRole.roleName;
    this.roleId = adminRole.roleId;
  }
}

export interface BillingRoleProps {
  /**
   * The name of the IAM policy for the IAM role.
   *
   * @default 'billing-role-s3olap-policy'
   */
  readonly policyName?: string;
  /**
   * The name of the object Lambda access point, which will be the same as the S3 acceess point for the S3 bucket in the demostration.
   *
   * @default 'billing-s3olap-call-transcripts-known-pii'
   */
  readonly objectLambdaAccessPointName?: string;
  /**
   * The name of the IAM role.
   *
   * @default 'RedactionBillingRole'
   */
  readonly iamRoleName?: string;
}

export class BillingRole extends cdk.Construct {
  /**
   * The ARN of the IAM role.
   */
  public readonly roleArn: string;
  /**
   * The name of the IAM role.
   */
  public readonly roleName: string;
  /**
   * The unique string identifying the role.
   */
  public readonly roleId: string;
  constructor(scope: cdk.Construct, id: string, props?: AdminRoleProps) {
    super(scope, id);
    const policyName = props?.policyName ?? 'billing-role-s3olap-policy';
    const objectLambdaAccessPointName = props?.objectLambdaAccessPointName ?? 'billing-s3olap-call-transcripts-known-pii';
    const iamRoleName = props?.iamRoleName ?? 'RedactionBillingRole';
    const billingRole = new iam.Role(this, 'RedactionBillingRole', {
      roleName: iamRoleName,
      assumedBy: new iam.AccountRootPrincipal(),
      inlinePolicies: {
        [policyName]: new iam.PolicyDocument({
          assignSids: true,
          statements: [
            new iam.PolicyStatement({
              sid: 'AllowListingObjects',
              effect: iam.Effect.ALLOW,
              actions: ['s3:ListBucket'],
              resources: ['*'],
            }),
            new iam.PolicyStatement({
              sid: 'AllowListingBucketsAndAccessPoints',
              effect: iam.Effect.ALLOW,
              actions: [
                's3:GetAccessPointForObjectLambda',
                's3:GetAccessPointConfigurationForObjectLambda',
                's3:ListAccessPointsForObjectLambda',
                's3:ListAllMyBuckets',
                's3:ListAccessPoints',
                's3:GetAccessPoint',
                's3:GetAccountPublicAccessBlock',
                's3:GetBucketPublicAccessBlock',
                's3:GetBucketPolicyStatus',
                's3:GetBucketAcl',
                's3:GetAccessPointPolicyStatus',
              ],
              resources: ['*'],
            }),
            new iam.PolicyStatement({
              sid: 'AllowObjectLambdaAccess',
              effect: iam.Effect.ALLOW,
              actions: [
                's3-object-lambda:Get*',
                's3-object-lambda:List*',
              ],
              resources: [`arn:${cdk.Aws.PARTITION}:s3-object-lambda:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:accesspoint/${objectLambdaAccessPointName}`],
            }),
            new iam.PolicyStatement({
              sid: 'AllowStandardAccessPointAccess',
              effect: iam.Effect.ALLOW,
              actions: [
                's3:Get*',
                's3:List*',
              ],
              resources: ['*'],
              conditions: {
                ['ForAnyValue:StringEquals']: {
                  'aws:CalledVia': ['s3-object-lambda.amazonaws.com'],
                },
              },
            }),
            new iam.PolicyStatement({
              sid: 'AllowLambdaInvocation',
              effect: iam.Effect.ALLOW,
              actions: [
                'lambda:InvokeFunction',
              ],
              resources: ['*'],
              conditions: {
                ['ForAnyValue:StringEquals']: {
                  'aws:CalledVia': ['s3-object-lambda.amazonaws.com'],
                },
              },
            }),
          ],
        }),
      },
    });
    this.roleArn = billingRole.roleArn;
    this.roleName = billingRole.roleName;
    this.roleId = billingRole.roleId;
  }
}

export interface CustSupportRoleProps {
  /**
   * The name of the IAM policy for the IAM role.
   *
   * @default 'customersupport-role-s3olap-policy'
   */
  readonly policyName?: string;
  /**
   * The name of the object Lambda access point, which will be the same as the S3 acceess point for the S3 bucket in the demostration.
   *
   * @default 'custsupport-s3olap-call-transcripts-known-pii'
   */
  readonly objectLambdaAccessPointName?: string;
  /**
   * The name of the IAM role.
   *
   * @default 'RedactionCustSupportRole'
   */
  readonly iamRoleName?: string;
}

export class CustSupportRole extends cdk.Construct {
  /**
   * The ARN of the IAM role.
   */
  public readonly roleArn: string;
  /**
   * The name of the IAM role.
   */
  public readonly roleName: string;
  /**
   * The unique string identifying the role.
   */
  public readonly roleId: string;
  constructor(scope: cdk.Construct, id: string, props?: AdminRoleProps) {
    super(scope, id);
    const policyName = props?.policyName ?? 'customersupport-role-s3olap-policy';
    const objectLambdaAccessPointName = props?.objectLambdaAccessPointName ?? 'custsupport-s3olap-call-transcripts-known-pii';
    const iamRoleName = props?.iamRoleName ?? 'RedactionCustSupportRole';
    const custSupportRole = new iam.Role(this, 'RedactionCustSupportRole', {
      roleName: iamRoleName,
      assumedBy: new iam.AccountRootPrincipal(),
      inlinePolicies: {
        [policyName]: new iam.PolicyDocument({
          assignSids: true,
          statements: [
            new iam.PolicyStatement({
              sid: 'AllowListingObjects',
              effect: iam.Effect.ALLOW,
              actions: ['s3:ListBucket'],
              resources: ['*'],
            }),
            new iam.PolicyStatement({
              sid: 'AllowListingBucketsAndAccessPoints',
              effect: iam.Effect.ALLOW,
              actions: [
                's3:GetAccessPointForObjectLambda',
                's3:GetAccessPointConfigurationForObjectLambda',
                's3:ListAccessPointsForObjectLambda',
                's3:ListAllMyBuckets',
                's3:ListAccessPoints',
                's3:GetAccessPoint',
                's3:GetAccountPublicAccessBlock',
                's3:GetBucketPublicAccessBlock',
                's3:GetBucketPolicyStatus',
                's3:GetBucketAcl',
                's3:GetAccessPointPolicyStatus',
              ],
              resources: ['*'],
            }),
            new iam.PolicyStatement({
              sid: 'AllowObjectLambdaAccess',
              effect: iam.Effect.ALLOW,
              actions: [
                's3-object-lambda:Get*',
                's3-object-lambda:List*',
              ],
              resources: [`arn:${cdk.Aws.PARTITION}:s3-object-lambda:${cdk.Aws.REGION}:${cdk.Aws.ACCOUNT_ID}:accesspoint/${objectLambdaAccessPointName}`],
            }),
            new iam.PolicyStatement({
              sid: 'AllowStandardAccessPointAccess',
              effect: iam.Effect.ALLOW,
              actions: [
                's3:Get*',
                's3:List*',
              ],
              resources: ['*'],
              conditions: {
                ['ForAnyValue:StringEquals']: {
                  'aws:CalledVia': ['s3-object-lambda.amazonaws.com'],
                },
              },
            }),
            new iam.PolicyStatement({
              sid: 'AllowLambdaInvocation',
              effect: iam.Effect.ALLOW,
              actions: [
                'lambda:InvokeFunction',
              ],
              resources: ['*'],
              conditions: {
                ['ForAnyValue:StringEquals']: {
                  'aws:CalledVia': ['s3-object-lambda.amazonaws.com'],
                },
              },
            }),
          ],
        }),
      },
    });
    this.roleArn = custSupportRole.roleArn;
    this.roleName = custSupportRole.roleName;
    this.roleId = custSupportRole.roleId;
  }
}