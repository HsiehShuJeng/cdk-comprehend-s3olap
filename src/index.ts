export * from './cdk-comprehend-s3olap';
export * from './comprehend-lambdas';
export * from './iam-roles';