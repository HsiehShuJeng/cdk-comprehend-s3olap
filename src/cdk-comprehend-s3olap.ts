import * as iam from '@aws-cdk/aws-iam';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';
// import * as s3objectlambda from '@aws-cdk/aws-s3objectlambda';
import * as comprehend_lambda from './comprehend-lambdas';
import * as s3loal_iam from './iam-roles';


export interface ComprehendS3olabProps extends comprehend_lambda.AccessConrtolLambdaProps, s3loal_iam.GeneralRoleProps {
  /**
     * The prefix attached to the name of the S3 bucket where you are going to explore the S3 Object Lambda.
     *
     * @default 6 random words
     */
  readonly s3BucketPrefix?: string;
  /**
     * The access point name for the access control Lambda with Amazon Comprehend.
     *
     * @default 'accessctl-s3-ap-survey-results-unknown-pii'
     */
  readonly accessControlAccessPointName?: string;
}

/**
 * Creates the foundation necessary to deploy the S3 Object Lambda Acceess Control Use Case.
 */
export class ComprehendS3olab extends cdk.Construct {
  // IAM role
  //   readonly policyName: string;
  //   readonly objectLambdaAccessPointName: string;
  //   readonly iamRoleName: string;
  // s3 bucket and access point
  //   readonly s3BucketPrefix: string;
  //   readonly accessControlAccessPointName: string;
  // object Lambda
  //   readonly semanticVersion: string;
  //   readonly confidenceThreshold: string;
  //   readonly containsPiiEntitiesThreadCount: string;
  //   readonly defaultLanguageCode: string;
  //   readonly documentMaxSize: string;
  //   readonly documentMaxSizeContainsPiiEntities: string;
  //   readonly isPartialObjectSupported: string;
  //   readonly logLevel: string;
  //   readonly maxCharsOverlap: string;
  //   readonly piiEntityTypes: string;
  //   readonly publishCloudWatchMetrics: string;
  //   readonly subsegmentOverlappingTokens: string;
  //   readonly unsupportedFileHandling: string;
  constructor(scope: cdk.Construct, id: string, props: ComprehendS3olabProps) {
    super(scope, id);
    // IAM role
    const policyName = props.policyName ?? 'general-role-s3olap-policy';
    const objectLambdaAccessPointName = props.objectLambdaAccessPointName ?? 'accessctl-s3olap-survey-results-unknown-pii';
    const iamRoleName = props.iamRoleName ?? 'GeneralRole';
    // s3 bucekt and its access point
    const s3BucketPrefix = props.s3BucketPrefix ?? this.generateS3Prefix(6);
    const accessControlAccessPointName = props.accessControlAccessPointName ?? 'accessctl-s3-ap-survey-results-unknown-pii';
    // object Lambda
    const semanticVersion = props.semanticVersion ?? '1.0.2';
    const confidenceThreshold = props.confidenceThreshold ?? '0.5';
    const containsPiiEntitiesThreadCount = props.containsPiiEntitiesThreadCount ?? '20';
    const defaultLanguageCode = props.defaultLanguageCode ?? 'en';
    const documentMaxSize = props.documentMaxSize ?? '102400';
    const documentMaxSizeContainsPiiEntities = props.documentMaxSizeContainsPiiEntities ?? '50000';
    const isPartialObjectSupported = props.isPartialObjectSupported ?? 'false';
    const logLevel = props.logLevel ?? 'INFO';
    const maxCharsOverlap = props.maxCharsOverlap ?? '200';
    const piiEntityTypes = props.piiEntityTypes ?? 'ALL';
    const publishCloudWatchMetrics = props.publishCloudWatchMetrics ?? 'true';
    const subsegmentOverlappingTokens = props.subsegmentOverlappingTokens ?? '20';
    const unsupportedFileHandling = props.unsupportedFileHandling ?? 'FAIL';

    const accessControlLambda = new comprehend_lambda.AccessConrtolLambda(this, 'ServerlessLambda', {
      semanticVersion: semanticVersion,
      confidenceThreshold: confidenceThreshold,
      containsPiiEntitiesThreadCount: containsPiiEntitiesThreadCount,
      defaultLanguageCode: defaultLanguageCode,
      documentMaxSize: documentMaxSize,
      documentMaxSizeContainsPiiEntities: documentMaxSizeContainsPiiEntities,
      isPartialObjectSupported: isPartialObjectSupported,
      logLevel: logLevel,
      maxCharsOverlap: maxCharsOverlap,
      piiEntityTypes: piiEntityTypes,
      publishCloudWatchMetrics: publishCloudWatchMetrics,
      subsegmentOverlappingTokens: subsegmentOverlappingTokens,
      unsupportedFileHandling: unsupportedFileHandling,
    });

    new s3loal_iam.GeneralRole(this, 'General', {
      objectLambdaAccessPointName: objectLambdaAccessPointName,
      policyName: policyName,
      iamRoleName: iamRoleName,
    });

    const surveyBucket = new s3.Bucket(this, 'SurveyResultBucket', {
      bucketName: `survey-results-unknown-pii-${s3BucketPrefix}`,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });
    surveyBucket.addToResourcePolicy(new iam.PolicyStatement(
      {
        sid: 'AWSBucketGetPolicy',
        principals: [new iam.AccountRootPrincipal()],
        actions: ['s3:GetObject'],
        effect: iam.Effect.ALLOW,
        resources: [`${surveyBucket.bucketArn}/*`],
        conditions: { ['StringEquals']: { 's3:DataAccessPointAccount': cdk.Aws.ACCOUNT_ID } },
      },
    ));
    surveyBucket.addToResourcePolicy(new iam.PolicyStatement(
      {
        sid: 'AWSBucketPutPolicy',
        principals: [new iam.AccountRootPrincipal()],
        actions: ['s3:PutObject'],
        effect: iam.Effect.ALLOW,
        resources: [`${surveyBucket.bucketArn}/*`],
      },
    ));
    surveyBucket.node.addDependency(accessControlLambda);
    // surveyBucket.node.addDependency(generalRole);

    new s3.CfnAccessPoint(this, 'AccessControlAccessPoint', {
      bucket: surveyBucket.bucketName,
      name: accessControlAccessPointName,
    });

    // new s3objectlambda.CfnAccessPoint(this, 'LambdaAccessPoint', {
    //     name: accessControlAccessPointName,
    //     objectLambdaConfiguration: {
    //         supportingAccessPoint: surveyAccessPoint.name,
    //         transformationConfigurations: [
    //             {
    //                 contentTransformation: JSON.stringify({
    //                     FunctionArn:
    //     })
    //             }
    //         ]
    //     }
    // });
  }

  private generateS3Prefix(length: number): string {
    var result = '';
    const characters = 'abcdefghijklmnopqrstuvwxyz0123456789';
    const charactersLength = characters.length;
    for (var i = 0; i < length; i++) {
      result += characters.charAt(Math.floor(Math.random() * charactersLength));
    }
    return result;
  }
}