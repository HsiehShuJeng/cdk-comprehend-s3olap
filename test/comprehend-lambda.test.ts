import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { AccessConrtolLambda } from '../src/comprehend-lambdas';

test('Comprehend Lambda test', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'ker-ker');
  new AccessConrtolLambda(stack, 'AccessControlLambda', {});
  const template = Template.fromStack(stack);

  template.resourceCountIs('AWS::Serverless::Application', 1);
  template.hasResourceProperties('AWS::Serverless::Application', {
    Location: {
      ApplicationId: 'arn:aws:serverlessrepo:us-east-1:839782855223:applications/ComprehendPiiAccessControlS3ObjectLambda',
      SemanticVersion: '1.0.2',
    },
    Parameters: {
      ConfidenceThreshold: '0.5',
      ContainsPiiEntitiesThreadCount: '20',
      DefaultLanguageCode: 'en',
      DocumentMaxSize: '102400',
      DocumentMaxSizeContainsPiiEntities: '50000',
      IsPartialObjectSupported: 'false',
      LogLevel: 'INFO',
      MaxCharsOverlap: '200',
      PiiEntityTypes: 'ALL',
      PublishCloudWatchMetrics: 'true',
      SubsegmentOverlappingTokens: '20',
      UnsupportedFileHandling: 'FAIL',
    },
  });
});