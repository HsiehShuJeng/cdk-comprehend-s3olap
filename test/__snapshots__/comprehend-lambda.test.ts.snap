// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comprehend Lambda test 1`] = `
Object {
  "Resources": Object {
    "AccessControlLambdaComprehendPiiAccessControlS3ObjectLambda22CE3759": Object {
      "Properties": Object {
        "Location": Object {
          "ApplicationId": "arn:aws:serverlessrepo:us-east-1:839782855223:applications/ComprehendPiiAccessControlS3ObjectLambda",
          "SemanticVersion": "1.0.2",
        },
        "Parameters": Object {
          "ConfidenceThreshold": "0.5",
          "ContainsPiiEntitiesThreadCount": "20",
          "DefaultLanguageCode": "en",
          "DocumentMaxSize": "102400",
          "DocumentMaxSizeContainsPiiEntities": "50000",
          "IsPartialObjectSupported": "false",
          "LogLevel": "INFO",
          "MaxCharsOverlap": "200",
          "PiiEntityTypes": "ALL",
          "PublishCloudWatchMetrics": "true",
          "SubsegmentOverlappingTokens": "20",
          "UnsupportedFileHandling": "FAIL",
        },
      },
      "Type": "AWS::Serverless::Application",
    },
  },
  "Transform": "AWS::Serverless-2016-10-31",
}
`;
